src/PriceOracle.ts

```diff
@@ -0,0 +1,138 @@
+    const rpcURL = CHAIN_CONSTANTS[chainId].rpcURL;
+    const web3 = new Web3(rpcURL);
+    const contract = new web3.eth.Contract(contractABI, contractAddress);
+
+    const numAddrs = addrs.length - 1;
+    const prices: string[] = await contract.methods.getManyRatesWithConnectors(numAddrs, addrs)
+        .call({}, blockNumber);
+    return prices;
+}
+
+/**
+ * Fetches the prices of whitelisted tokens for a given blockchain network.
+ *
+ * This function retrieves the list of whitelisted tokens based on the provided
+ * chain ID, fetches their current prices from a price oracle, and updates the
+ * context with token information including their addresses, symbols, units, and prices.
+ *
+ * @param {number} chainId - The ID of the blockchain network to fetch prices for.
+ *                           Use 10 for the Optimism network, or any other value
+ *                           for the base network.
+ * @param {number} blockNumber - The block number to fetch prices for.
+ * @param {Date} blockDatetime - The datetime of the block to use for updating timestamps.
+ * @param {any} context - The context object to interact with the Token and TokenPrice entities.
+ * @returns {Promise<void>} A promise that resolves when the operation is complete.
+ *
+ * @throws {Error} Throws an error if the price fetching process fails.
+ */
+export async function set_whitelisted_prices(chainId: number, blockNumber: number, blockDatetime: Date, context: any): Promise<void> {
+    // Skip if not yet available
+    let startBlock = CHAIN_CONSTANTS[chainId].oracle.startBlock || Number.MAX_SAFE_INTEGER;
+    if (blockNumber < startBlock) return;
+
+    const lastUpdated = getPricesLastUpdated(chainId);
+    const timeDelta = CHAIN_CONSTANTS[chainId].oracle.updateDelta * 1000;
+    const tokensNeedUpdate = !lastUpdated || (blockDatetime.getTime() - lastUpdated.getTime()) > timeDelta;
+    if (!tokensNeedUpdate) return;
+
+    // Get token data for chain
+    const tokenData = chainId === 10 ? OPTIMISM_WHITELISTED_TOKENS : BASE_WHITELISTED_TOKENS;
+
+    // Get prices from oracle
+    const addresses = tokenData
+        .filter(token => token.createdBlock <= blockNumber)
+        .map(token => token.address);
+
+    if (addresses.length === 0) return;
+    const units = tokenData.map(token => token.unit as validUnit);
+    const prices = await read_prices(addresses, chainId, blockNumber);
+
+    // Map prices to token addresses
+    const pricesByAddress = new Map<string, number>();
+    pricesByAddress.set(CHAIN_CONSTANTS[chainId].usdc.address, 1);
+
+    prices.forEach((price, index) => {
+        pricesByAddress.set(addresses[index], Number(utils.fromWei(price, units[index])));
+    });
+    
+    for (const token of tokenData) {
+        const price = pricesByAddress.get(token.address) || 0;
+        
+        // Get or create Token entity
+        let tokenEntity = await context.Token.get(token.address);
```

Comment:
Heads up reading from db will decrease speed slightly but this is only really here and there so should be fine.